src/main/java/carrierManagementSystem/CarrierManagementSystemDirector.java

```diff
@@ -0,0 +1,74 @@
+package carrierManagementSystem;
+
+import java.util.Stack;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.EmptyStackException;
+import java.util.List;
+
+/**
+ * This singleton class is responsible to create LTE carrier and WCDMA carrier. 
+ * We can call this class director as it will direct the creation command from
+ * client(CarrierManagementSystemClient) to Carrier builder class. This class
+ * will also assign a random unique carrier ID (in range 1 - 1000) to each
+ * carrier it creates.
+ */
+public class CarrierManagementSystemDirector implements CarrierManagementIf {
```

Comment:
The two methods in this class `createLteCarrier()` and `createWcdmaCarrier()` are almost identical - the two main differences between them are the use of which builder instantiated and used (`LteCarrierBuilder` and `WcdmaCarrierBuilder` classes) and the method we call on that class (`getLteCarrier()` and `getWcdmaCarrier()`, which could be named more generically and put in the `CarrierBuilder` interface). Since both classes implement the `CarrierBuilder` interface (which, as a side note, isn't referenced anywhere), could some refactoring be done?
In my mind, this class would have a member variable of type `CarrierBuilder` and each of the public methods (`createLteCarrier` and `createWcdmaCarrier`) would set that member variable to the appropriate concrete class, then call a common method that calls each of the steps for creating the `Carrier`. This would also involve adding a public method to the `CarrierBuilder` class called `getCarrier()`, that could also be called in that common method. Let me know if you'd like me to explain this further (since this comment is getting rather long, and might not aptly describing what it is intended to)

Reply:
we may need to talk more about it  however I made some changes into interface to make it more generic.